### PR TITLE
[XWI-133] add queries to fetch page hierachy information

### PR DESCRIPTION
**Pull request by Kharonus**
# Ticket
[XWI-133](https://community.openproject.org/wp/XWI-133)

# What approach did you choose and why?
- add two internal queries
- add contracts, input and result model
- HINT: this PR is a first step of https://community.openproject.org/wp/XWI-107
- next step is amending the search service

### CODE REVIEW

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Wikis
+  module Adapters
+    module Providers
+      module Internal
+        module Queries
+          module Internal
```
Comment (Kharonus) [inline modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: I hate it. And here I give the 🎤 to @NobodysNightmare  and he can tell me: "Told you so."

What is happening? I voted to use a namespace called "internal" for queries and commands, that are not exposed via the registry. But I completely overlooked, that we already have an `internal` namespace for the provider type.

So, I'm ready to revisit that decision. WDYT? Shall I move all (2) queries out of this namespace and basically we rely that not exposing those query via the registry is the only indicator, that those are considered for "internal" use only?

**Comment by Kharonus** [comment]
Missing tests ...

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 1 above)
Comment (NobodysNightmare) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: I'd be fine to move internal queries out of the `Internal` namespace to avoid confusion.

Comment (NobodysNightmare) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: Mh. Wait a minute. This PR exposes a problem with PRs that are too small: I can't see how those queries are intended to be used, there are no callers to them.

And this is kinda crucial here, because I am starting to get confused. Apparently you are planning for the `PageHierarchy` query to be internal. That would mean they are only used "inside" the provider. On the other hand, I know that they will be used for a feature that works across providers.

To me those two things don't line up yet. Either an outside caller wants to call those queries (making them non-internal) **or** another query wants to use them.

Do you have a preview of the PR that intends to use the queries somewhere already?

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+              class << self
+                def wiki_page_to_page_hierarchy(wiki_page, provider:)
+                  Results::PageHierarchy.new(
+                    wiki: wiki_result_from_wiki(wiki_page.wiki, provider:),
+                    page: PageInfo.wiki_page_to_page_info(wiki_page, provider:),
+                    ancestors: wiki_page.ancestors.map { PageInfo.wiki_page_to_page_info(it, provider:) }
+                  )
+                end
+
+                def wiki_result_from_wiki(wiki, provider:)
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: 🟡 This method name is ambiguous, though I understand why you chose it.

The thing is that in other places we talk about a "page info **result**", when we don't return a `PageInfo`, but a `Dry::Result(PageInfo)`. It is a hint at the fact that you don't get the thing directly, but there is a result to unpack first.

This is not the case here. Here we wanted to clarify that you don't get a `Wiki` (the AR model)`, but a `...::Results::Wiki`. Short: a wiki result.

It would also sound weird to call this method `wiki_from_wiki`.

Another thing I notice is that the other methods are called `*_to_*`, but this one is called `*_from_*`.

So yeah. I don't have good ideas, but take these bad ones:

* `ar_wiki_to_wiki`
* `activerecord_wiki_to_wiki`
* `wiki_model_to_wiki`
* `wiki_to_adapter_wiki`

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                  Results::Wiki.new(
+                    identifier: wiki.id.to_s,
+                    provider:,
+                    name: wiki.project.name,
+                    # FIXME: internal wiki currently has no notion of a single root page, nor an official "entrypoint"
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: Alternatively we could link to the project's overview page. We take its name as well, so why not also taking its href?

In many cases the "first main page" will already be linked in the hierarchy anyways (one level deeper).

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                delegate :url_for, to: :url_helpers
+
+                def url_helpers
+                  @url_helpers ||= OpenProject::StaticRouting::StaticRouter.new.url_helpers
+                end
+              end
+
+              def call(input_data:, auth_strategy:)
+                Adapters::Authentication[auth_strategy].call do |user|
+                  wiki_page = WikiPage.visible(user).find_by(id: input_data.identifier)
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: 🟡  Mh. Do we have any chance to `include(:ancestors)` or will this not fly, because that's a recursive relation?

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb [hunk 5]
```diff
@@ -53,6 +53,10 @@ module ClassMethods
               def fetch_json(json_hash, key)
                 json_hash.fetch(key) { throw :xwiki_error, Failure(Results::Error.new(source: self, code: :invalid_response)) }
               end
+
+              def dig_json(json_hash, *keys)
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb]: ❗ Uh this is interesting. In a very recent PR of mine (https://github.com/opf/openproject/pull/24049) I decided against the usage of `dig`, because it didn't behave the way I wanted it to.

I am leaving this here for you to decide, but hear me out :D

Essentially `dig` looks nice to go into unknown data at first, because it works well with the expected data as well as with unexpected data and it even mixes arrays and hashes:

```ruby
{ a: [{ c: 42 }] }.dig(:a, 0, :c)
=> 42
{ a: [{ c: 42 }] }.dig(:a, 100, :c)
=> nil
{ a: [{ c: 42 }] }.dig(:a, 0, :d)
=> nil
```

However, as soon as the structure mismatches in unexpected ways, you start receiving weird errors:

```ruby
{ a: [{ c: 42 }] }.dig(:a, :b, :c)
(irb):1:in 'Hash#dig': no implicit conversion of Symbol into Integer (TypeError)
```

This means error handling gets weird. You have to react to both `nil` responses **and** exceptions. That of course combines with the usual weakness of not being able to distinguish a "real" `nil` and a missing value.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                    provider:,
+                    name: dig_json(data, "hierarchy", "items", 0, "label"),
+                    href: dig_json(data, "hierarchy", "items", 0, "url")
+                  )
+                end
+
+                def json_to_ancestors(data, provider:)
+                  page_title = fetch_json(data, "title")
+
+                  invalid_hierarchy_types = %w[document wiki]
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: What are the _valid_ hierarchy types? Should we maybe use an allow list here?

Since we only expect pages/spaces, I can't imagine that it's an open list of valid types.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                  ancestors = dig_json(data, "hierarchy", "items").filter do |item|
+                    next false if invalid_hierarchy_types.include?(fetch_json(item, "type"))
+
+                    fetch_json(item, "label") != page_title
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: 🔴 Is that a valid condition?

What if I have a hierarchy like

* Overview
    * Rooms
        * Storage Room
            * Overview

This hierarchy now contains "Overview" twice. Of course that's not great, but I also wouldn't know why we should prevent it.

Is this to filter the reference to the page itself? Couldn't this be solved by dropping the first/last element of the hierarchy? (probably both, because we filter wikis as well ^^)

**Review by NobodysNightmare — CHANGES_REQUESTED** [review]

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 1 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: No, those are indeed internal. They will be used from the `search_pages` query. When searching, we no longer return a list of page infos, but a list of page hierarchies then.

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 2 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: Yes, I agree. Let me choose a better name

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 3 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: I agree, in fact, the URL is not used yet anywhere. And we did say, that we want to encourage the understanding of "the project is the wiki".

will change

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 4 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: it's recursive, if you execute that relation you see multiple SQL statements in the logs, all the same fetching the next parent.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb (hunk 5 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb]: I agree, I can make a more sophisticated implementation of the base class method.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb (hunk 6 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: hmm, I'm only interested in `type == space` here, so I can swap the expectation. This grew from slightly different code.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb (hunk 7 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: Ah, crap.... this is bad, because right now I only have the name. Yeah, I need the identifier here, too.

But no, the reason why I exclude ancestors with the same title as the page, is that the `WebHome` appears in the ancestors, too, and it has the same label as the "leaf".

Comment (NobodysNightmare) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: All `WebHome`? I am still wondering whether we can (safely and reliably) do things like "skip last element" or "skip first element".

But probably not, because a "Leaf Space" would have a `WebHome` and a "Leaf Final Page" would not, right?

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 2 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: Hell, I chose one of the bad ones :D

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 1 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: Moved out of internal namespace

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb (hunk 5 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/concerns/xwiki_request.rb]: took your approach and ported it to xwiki requests concern.

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb (hunk 7 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: exactly, I found both examples. :/

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb (hunk 6 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: changed

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/canonical_page_hierarchy.rb [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+                  name: fetch_json(data, "hierarchy", "items", 0, "label"),
+                  href: fetch_json(data, "hierarchy", "items", 0, "url")
+                )
+              end
+
+              def json_to_ancestors(data, provider:)
+                page_title = fetch_json(data, "title")
+
+                ancestors = fetch_json(data, "hierarchy", "items").filter do |item|
+                  next false unless fetch_json(item, "type") == "space"
```
Comment (NobodysNightmare) [inline modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/canonical_page_hierarchy.rb]: 🟢 only a nit:


```suggestion
                  next false if fetch_json(item, "type") != "space"
```

File: modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb (hunk 3 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/internal/queries/internal/page_hierarchy.rb]: changed

File: modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb (hunk 7 above)
Comment (Kharonus) [reply modules/wikis/app/services/wikis/adapters/providers/xwiki/queries/internal/canonical_page_hierarchy.rb]: no longer needed, hierarchy is specifically fetched from endpoint

File: modules/wikis/spec/services/wikis/adapters/providers/internal/queries/page_hierarchy_query_spec.rb [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      let(:identifier) { "THIS IS NO MOON!" }
+
+      it { is_expected.to be_failure }
+    end
+
+    context "when user can't see wiki page" do
+      let(:wiki_project_permissions) { %i[view_work_packages] }
+
+      it { is_expected.to be_failure }
+    end
```
Comment (NobodysNightmare) [inline modules/wikis/spec/services/wikis/adapters/providers/internal/queries/page_hierarchy_query_spec.rb:123]: 🟢 Probably not a crazily broken case, but I'd also include a context "when wiki page has no ancestors", because such edge cases tend to attract bugs.

Take this as a nit though, existing tests already look nice.

**Review by NobodysNightmare — APPROVED** [review]
LGTM.

I am surprised how much simpler the XWiki implementation looks now. Did the actual _list_ of ancestors change with the latest API changes on the XWiki side? I remember us talking about having to filter out "WebHome". This doesn't seem to be a topic anymore.

File: modules/wikis/spec/services/wikis/adapters/providers/internal/queries/page_hierarchy_query_spec.rb (hunk 9 above)
Comment (Kharonus) [reply modules/wikis/spec/services/wikis/adapters/providers/internal/queries/page_hierarchy_query_spec.rb:123]: Will add, easy enough.

**Comment by Kharonus** [comment]
It is a fully different endpoint now, it returns the same objects - with every nested information - for each ancestor. So, the response is basically an array of page responses.
